presto-main/src/main/java/com/facebook/presto/operator/window/WindowPartition.java

```diff
@@ -189,16 +256,229 @@ else if (frameInfo.getEndType() == PRECEDING) {
         else if (frameInfo.getEndType() == FOLLOWING) {
             frameEnd = following(rowPosition, endPosition, getEndValue(frameInfo));
         }
-        else if (frameInfo.getType() == RANGE) {
+        else {
+            frameEnd = rowPosition;
+        }
+
+        return new Range(frameStart, frameEnd);
+    }
+
+    private Range getFrameRange(FrameInfo frameInfo, Range recentRange, PagesIndexComparator startComparator, PagesIndexComparator endComparator)
+    {
+        // full partition
+        if ((frameInfo.getStartType() == UNBOUNDED_PRECEDING && frameInfo.getEndType() == UNBOUNDED_FOLLOWING)) {
+            return new Range(0, partitionEnd - partitionStart - 1);
+        }
+
+        // frame defined by peer group
+        if ((frameInfo.getStartType() == CURRENT_ROW && frameInfo.getEndType() == CURRENT_ROW) ||
+                (frameInfo.getStartType() == CURRENT_ROW && frameInfo.getEndType() == UNBOUNDED_FOLLOWING) ||
+                (frameInfo.getStartType() == UNBOUNDED_PRECEDING && frameInfo.getEndType() == CURRENT_ROW)) {
+            // same peer group as recent row
+            if (currentPosition == partitionStart || pagesIndex.positionEqualsPosition(peerGroupHashStrategy, currentPosition - 1, currentPosition)) {
+                return recentRange;
+            }
+            // next peer group
+            return new Range(
+                    frameInfo.getStartType() == UNBOUNDED_PRECEDING ? 0 : peerGroupStart - partitionStart,
+                    frameInfo.getEndType() == UNBOUNDED_FOLLOWING ? partitionEnd - partitionStart - 1 : peerGroupEnd - partitionStart - 1);
+        }
+
+        // at this point, frame definition has at least one of: X PRECEDING, Y FOLLOWING
+        // 1. leading or trailing nulls: frame consists of nulls peer group, possibly extended to partition start / end.
+        // according to Spec, behavior of "X PRECEDING", "X FOLLOWING" frame boundaries is similar to "CURRENT ROW" for null values.
+        if (pagesIndex.isNull(frameInfo.getSortKeyChannel(), currentPosition)) {
+            return new Range(
+                    frameInfo.getStartType() == UNBOUNDED_PRECEDING ? 0 : peerGroupStart - partitionStart,
+                    frameInfo.getEndType() == UNBOUNDED_FOLLOWING ? partitionEnd - partitionStart - 1 : peerGroupEnd - partitionStart - 1);
+        }
+
+        // 2. non-null value in current row. Find frame boundaries starting from recentRange
+        int frameStart;
+        if (frameInfo.getStartType() == UNBOUNDED_PRECEDING) {
```

Comment:
can you use case statements here? would be clearer.  also, maybe throw an exception if startType isn't any of these